external/concealer/include/concealer/utility/service_with_validation.h

```diff
@@ -0,0 +1,97 @@
+// Copyright 2015 TIER IV, Inc. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef CONCEALER__SERVICE_WITH_VALIDATION_HPP_
+#define CONCEALER__SERVICE_WITH_VALIDATION_HPP_
+
+#include <chrono>
+#include <string>
+
+#include "concealer/autoware.hpp"
+#include "rclcpp/rclcpp.hpp"
+
+namespace concealer
+{
+template <class T>
+class ServiceWithValidation
+{
+public:
+  ServiceWithValidation(std::string serviceName, Autoware & autoware)
+  : serviceName(serviceName),
+    logger(autoware.get_logger()),
+    client(autoware.create_client<T>(serviceName, rmw_qos_profile_default)),
+    validation_rate(std::chrono::seconds(1))
+  {
+  }
+
+  void operator()(const typename T::Request::SharedPtr & request)
+  {
+    validateAvailability();
+    callWithTimeoutValidation(request);
+  }
+
+  void operator()(const typename T::Request::SharedPtr & request, size_t attempts_count)
+  {
+    validateAvailability();
+    for (size_t attempt = 0; attempt < attempts_count; ++attempt, validation_rate.sleep()) {
+      std::optional<typename rclcpp::Client<T>::SharedFuture> future_opt =
+        callWithTimeoutValidation(request);
+      if (!future_opt) {
+        continue;
+      }
+
+      auto result = future_opt->get();
+      if (result->status.code == tier4_external_api_msgs::msg::ResponseStatus::SUCCESS) {
+        RCLCPP_INFO_STREAM(
+          logger, serviceName << " service request has been accepted "
+                              << (result->status.message.empty()
+                                    ? "."
+                                    : " (" + result.get()->status.message + ")."));
+        break;
```

Comment:
Rather than writing a process with an early return using break, the branch should be written carefully using the else clause.

Reply:
I am sorry but I am not sure how to implement this. Could you elaborate on what would be the desirable code structure? :bow:

Reply:
I think loops with early return, continue, and break are prone to mistakes, so I think the pattern should be clearly shown by the exit condition of the for loop and the branch with the else clause.

For example, the break on line 61 in your code effectively leads to an exit from the function, but it's hard to read the intent at first glance.
Also, an error should be thrown if the loop completes attempts_count times and still does not succeed, but if you add that processing to the end of this function, break will also throw an error on success.

However, when I reviewed the code again, the part "the branch should be written carefully using the else clause" was incorrectly pointed out. I'm sorry.

After all, I think this function should be rewritten as:

```
auto operator()(const typename T::Request::SharedPtr & request, std::size_t attempts_count) -> void
{
  validateAvailability();

  for (auto attempt = attempts_count; 0 < attempt; --attempt)
  {
    if (auto future_opt = callWithTimeoutValidation(request);
        future_opt and
        future_opt->get()->status.code == tier4_external_api_msgs::msg::ResponseStatus::SUCCESS)
    {
      RCLCPP_INFO_STREAM(
        logger, serviceName << " service request has been accepted "
                            << (future_opt->get()->status.message.empty()
                                  ? "."
                                  : " (" + future_opt->get().get()->status.message + ")."));
      return;
    }
    else
    {
      RCLCPP_ERROR_STREAM(
        logger,
        serviceName << " service request was accepted, but ResponseStatus is FAILURE "
                    << (future_opt->get()->status.message.empty() ? "" : " (" + future_opt->get()->status.message + ")"));

      validation_rate.sleep();
    }
  }

  // THROW ERROR "requested the service attempts_count times, but was not successful."
}
```

Reply:
After rewriting it, I felt that it is difficult to implement this function cleanly.
I don't think the code I showed above is pretty either.

Reply:
Thank you for the clarification. I have applied the return within the loop and error reporting, but I left the second if statement with the initializer of `service_call_status`, to avoid the repetition of `service_call_result->get()->status`.

Would the current structure be ok?

Reply:
> Would the current structure be ok?
ok!